### PR TITLE
Unregister onFrameAvailable callbacks when a TextureEntry is released

### PR DESCRIPTION
**Pull request by nichtverstehen**
Otherwise the callbacks may be called after FlutterNativeView is destroyed and is null.

Also defensively check for whether the texture is already released in the callback because the callback may be called from another thread by a stale reference (see the comment).

This closes https://github.com/flutter/flutter/issues/20951.

### CODE REVIEW

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 1]
```diff
@@ -997,6 +997,14 @@ public void onReceive(Context context, Intent intent) {
             this.surfaceTexture.setOnFrameAvailableListener(new SurfaceTexture.OnFrameAvailableListener() {
                 @Override
                 public void onFrameAvailable(SurfaceTexture texture) {
+                    // NB: This callback may be called on an arbitrary thread.
+                    if (SurfaceTextureRegistryEntry.this.surfaceTexture.isReleased()) { // isReleased is synchronized in JNI code.
```
Comment (mehmetf) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Android BUILD failed.

Unfortunately, this API only seems to be available in Android 26 and up. You will need to do a SDK version check here too :-/

Comment (nichtverstehen) [reply shell/platform/android/io/flutter/view/FlutterView.java]: Right. I also realized that it's still racy: if all of the deinitialization happens between checking isReleased() and accessing mNativeView we'd get null anyway.

We really should pass a Handler and execute the callback on the main thread. Is nativeMarkTextureFrameAvailable actually meant to be thread-safe?

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 2]
```diff
@@ -997,9 +997,16 @@ public void onReceive(Context context, Intent intent) {
...
+                        // Even though we make sure to unregister the callback before releasing,
+                        // SurfaceTexture has a data race when accessing the callback, so the callback may
+                        // still be called by a stale reference after released==true and mNativeView==null.
+                        // See b/113099095.
+                        return;
+                    }
                     nativeMarkTextureFrameAvailable(mNativeView.get(), SurfaceTextureRegistryEntry.this.id);
                 }
-            });
+            }, new Handler());
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Why the `new Handler()`?

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 3]
```diff
@@ -997,9 +997,16 @@ public void onReceive(Context context, Intent intent) {
             this.surfaceTexture.setOnFrameAvailableListener(new SurfaceTexture.OnFrameAvailableListener() {
                 @Override
                 public void onFrameAvailable(SurfaceTexture texture) {
+                    if (released) {
+                        // Even though we make sure to unregister the callback before releasing,
+                        // SurfaceTexture has a data race when accessing the callback, so the callback may
+                        // still be called by a stale reference after released==true and mNativeView==null.
+                        // See b/113099095.
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Can we avoid referencing buganizer from the framework?

Comment (matthew-carroll) [reply shell/platform/android/io/flutter/view/FlutterView.java]: If necessary, please copy the buganizer bug over to GitHub and then reference that here instead.

**Comment by matthew-carroll** [comment]
The changes look fine to me.  I also added Chinmay and Jason for a sanity check.  I'm not familiar enough with FlutterView and FlutterNativeView yet to be sure there aren't any implications to this.

File: shell/platform/android/io/flutter/view/FlutterView.java (hunk 3 above)
Comment (nichtverstehen) [reply shell/platform/android/io/flutter/view/FlutterView.java]: Removed the bug reference. (FWIW, the buganizer bug is against Android base system).

File: shell/platform/android/io/flutter/view/FlutterView.java (hunk 2 above)
Comment (nichtverstehen) [reply shell/platform/android/io/flutter/view/FlutterView.java]: Added a comment. See also the discussion with Mehmet in the comments above.

**Comment by nichtverstehen** [comment]
Folks, what can we do to get this merged?

**Comment by matthew-carroll** [comment]
@mehmetf Do one of us need to merge this in?

**Pull request merged**